pilot/pkg/config/memory/monitor.go

```diff
@@ -72,32 +79,43 @@ func newBufferedMonitor(store model.ConfigStore, bufferSize int, sync bool) Moni
 		store:    store,
 		handlers: handlers,
 		eventCh:  make(chan ConfigEvent, bufferSize),
-		sync:     sync,
+		sync:     syncMode,
+		lock:     &sync.RWMutex{},
 	}
 }
 
 func (m *configStoreMonitor) ScheduleProcessEvent(configEvent ConfigEvent) {
 	if m.sync {
 		m.processConfigEvent(configEvent)
-	} else {
-		m.eventCh <- configEvent
+		return
+	}
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.closed.Load() {
+		return
 	}
+	m.eventCh <- configEvent
+}
+
+func (m *configStoreMonitor) waitAndCleanUp(stop <-chan struct{}) {
+	<-stop
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.closed.Store(true)
+	// indicating the monitor has exited.
+	close(m.eventCh)
 }
 
 func (m *configStoreMonitor) Run(stop <-chan struct{}) {
 	if m.sync {
 		<-stop
 		return
 	}
-	for {
-		select {
-		case <-stop:
-			return
-		case ce, ok := <-m.eventCh:
-			if ok {
-				m.processConfigEvent(ce)
-			}
-		}
+
+	go m.waitAndCleanUp(stop)
```

Comment:
I doubt this is not what we want. You must call Run and wait it exit in same thread.

But I think we need async notification to let the caller to know it has processed all events asynchronously

Reply:
Ok，what about adding new function to the [monitor interface](https://github.com/istio/istio/blob/890986aeae87fbe211c3a7986cc7ad272217bb1d/pilot/pkg/config/memory/monitor.go#L32),  
```
// Monitor provides methods of manipulating changes in the config store
type Monitor interface {
	Run(<-chan struct{})
	AppendEventHandler(config2.GroupVersionKind, Handler)
	ScheduleProcessEvent(ConfigEvent)
        // HasCompleted returns true if all the ConfigEvents  has been
	// processed in async mode.  Returns true directly in sync mode.
        HasCompleted() bool
}
```